File: job11/src/pfc.py
def produce(te):
    print("PRODUCE")

    i=0
    while(i<8):
        print(f"produce {i}")
        yield i
        i=i+1

def filter(te):
    print("FILTER")

    i=0
    while(i<8):
        x=next(coro_p)
        print(f"filter {i}->{i*10}")
        yield i*10
        i=i+1

def consume(te):
    print("\tCONSUME")

    i=0
    while(i<8):
        x = next(coro_f)
        print(f"\tconsume {i}")
        i=i+1

    print("END")

coro_p = produce(8)
coro_f =filter(8)
consume(8)

File: job11/src/pfc.c
#include <stdio.h>
#include <stdlib.h>
#include "lib/coro.h"

coro_t *coro_p;
coro_t *coro_c;
coro_t *coro_f;

void produce()
{
    puts("PRODUCE");

    for(int i=0;i<8;i++){
        printf("produce %d\n",i);
        coro_yield(i);
    }

    puts("NEVER REACH HERE");
}

void filter(){
    int x;
    puts("FILTER");

    for(int i=0;i<8;i++){
        x=coro_resume(coro_p);
        printf("filter %d->%d\n",x,x*10);
        coro_yield(i);
    }

}

void consume()
{
    int x;

    puts("\tCONSUME");

    for(int i=0;i<8;i++){
        x = coro_resume(coro_f);
        printf("\tconsume %d\n", x);
    }

    puts("END");
    exit(0);
}

int main()
{
    coro_p = coro_new(produce);
    coro_c = coro_new(consume);
    coro_f = coro_new(filter);
    coro_boot(coro_c);
    return 0;
}
